src/data/finset/noncomm_prod.lean

```diff
@@ -0,0 +1,169 @@
+/-
+Copyright (c) 2021 Yakov Pechersky. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yakov Pechersky
+-/
+
+import data.finset.fold
+
+/-!
+# Products (respectively, sums) over a finset or a multiset.
+
+The regular `finset.prod` and `multiset.prod` require `[comm_monoid α]`.
+Often, there are collections `s : finset α` where `[monoid α]` and we know,
+in a dependent fashion, that for all the terms `∀ (x ∈ s) (y ∈ s), commute x y`.
+This allows to still have a well-defined product over `s`.
+
+## Main definitions
+
+- `finset.noncomm_prod`, requiring a proof of commutativity of held terms
+- `multiset.noncomm_prod`, requiring a proof of commutativity of held terms
+
+## Implementation details
+
+While `list.prod` is defined via `list.foldl`, `noncomm_prod` is defined via
+`list.foldr` for neater proofs and definitions. By the commutativity assumption,
+the two must be equal.
+
+-/
+
+variables {α β : Type*} (f : α → β → β) (op : α → α → α)
+
+namespace multiset
+
+/-- Fold of a `s : multiset α` with `f : α → β → β`, given a proof that `left_commutative f`
+on all elements `x ∈ s`. -/
+def noncomm_foldr (s : multiset α) (comm) (b : β) : β :=
+s.attach.foldr (f ∘ subtype.val) comm b
```

Comment:
This definition might be more convenient if it handled the subtype recursor for the caller:
```suggestion
def noncomm_foldr (s : multiset α)
  (comm : ∀ (a₁ ∈ s) (a₂ ∈ s) b, f a₁ (f a₂ b) = f a₂ (f a₁ b)) (b : β) : β :=
s.attach.foldr (f ∘ subtype.val) (λ a₁ a₂, comm a₁ a₁.prop a₂ a₂.prop) b
```